File: examples/java-junit-gradle/src/test/java/examples/LibrarySteps.java
package examples;

import dev.varar.Registrar;
import dev.varar.State;
import dev.varar.StateBinder;
import dev.varar.StepDefinitions;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Locale;
import java.util.regex.Pattern;
import java.util.stream.Stream;

public final class LibrarySteps implements StepDefinitions {

    record Ctx(List<Library.Loan> loans, Library.Money fee, boolean granted) implements State {}

    /** June 6, 2026 ⇄ LocalDate 2026-06-06 — one formatter drives both parse and format. */
    private static final DateTimeFormatter DATE = DateTimeFormatter.ofPattern("MMMM d, yyyy", Locale.ENGLISH);

    /** £2.50 and 50p, both as GBP Money. */
    private static Library.Money toMoney(String raw) {
        return raw.endsWith("p")
                ? Library.gbp(Double.parseDouble(raw.substring(0, raw.length() - 1)) / 100)
                : Library.gbp(Double.parseDouble(raw.substring(1)));
    }

    /** The inverse: mismatches render as £2.60 / 50p, not as a Money dump. */
    private static String formatMoney(Library.Money m) {
        return m.value() < 1 ? Math.round(m.value() * 100) + "p" : String.format(Locale.ROOT, "£%.2f", m.value());
    }

    @Override
    public void defineSteps(Registrar registrar) {
        StateBinder<Ctx> s = registrar.steps(() -> new Ctx(List.of(), Library.gbp(0), false));

        s.param(
                "date",
                Pattern.compile("[A-Z][a-z]+ \\d{1,2}, \\d{4}"),
                groups -> LocalDate.parse(groups[0], DATE),
                DATE::format);
        // £2.50 and 50p, both as GBP Money. The amount is cucumber-expressions'
        // float regexp, minus the scientific notation.
        s.param(
                "money",
                Pattern.compile("£(?=.*\\d.*)[-+]?\\d*(?:\\.(?=\\d.*))?\\d*|\\d+p"),
                groups -> toMoney(groups[0]),
                LibrarySteps::formatMoney);
        // The emphasised run IS the parameter: the markers live in the pattern,
        // parse strips them, format restores them. Markup is notation, like £2.50.
        s.param(
                "title",
                Pattern.compile("\\*[^*]+\\*"),
                groups -> groups[0].substring(1, groups[0].length() - 1),
                title -> "*" + title + "*");

        s.stimulus(
                "borrowed {title}, due back on {date}",
                (Ctx ctx, String title, LocalDate due) -> new Ctx(
                        Stream.concat(ctx.loans().stream(), Stream.of(new Library.Loan(title, due)))
                                .toList(),
                        ctx.fee(),
                        ctx.granted()));

        s.stimulus(
                "returns it on {date}",
                (Ctx ctx, LocalDate returnedOn) -> new Ctx(
                        ctx.loans(),
                        ctx.loans().stream()
                                .map(loan -> Library.lateFee(loan, returnedOn))
                                .reduce(Library.gbp(0), Library::addMoney),
                        ctx.granted()));

        s.sensor("owes a {money} late fee", (Ctx ctx, Library.Money expected) -> ctx.fee());

        s.sensor("{money} for each day overdue", (Ctx ctx, Library.Money expected) -> Library.FEE_PER_DAY);

        s.stimulus(
                "asks to borrow {title} on {date}",
                (Ctx ctx, String title, LocalDate on) ->
                        new Ctx(ctx.loans(), ctx.fee(), Library.mayBorrow(ctx.loans(), on)));

        s.sensor("the library refuses", (Ctx ctx) -> {
            if (ctx.granted()) throw new AssertionError("expected the library to refuse");
            return null;
        });

        s.sensor("the library agrees", (Ctx ctx) -> {
            if (!ctx.granted()) throw new AssertionError("expected the library to agree");
            return null;
        });
    }
}

File: examples/typescript-vitest/steps/library.steps.ts
import { steps } from '@varar/varar'
import { addMoney, FEE_PER_DAY, GBP, type Loan, lateFee, type Money, mayBorrow } from './library'

// Custom parameter types are declared with chained `.param()` calls so their
// parse return types flow into the steps: {date} → Date, {money} → Money,
// {title} → string. The step handlers below need no argument annotations as a
// result.
const { stimulus, sensor } = steps(() => ({
  loans: [] as ReadonlyArray<Loan>,
  fee: GBP(0),
  granted: false,
}))
  // June 6, 2026 ⇄ a Date. (Temporal.PlainDate would fit better — swap it
  // in once Node ships Temporal unflagged; today it's behind --harmony-temporal.)
  .param(
    'date',
    /[A-Z][a-z]+ \d{1,2}, \d{4}/,
    (raw) => new Date(raw),
    (d) => d.toLocaleDateString('en-US', { dateStyle: 'long' }),
  )
  // £2.50 and 50p, both as GBP Money. The amount is cucumber-expressions'
  // float regexp, minus the scientific notation. The inverse format renders
  // mismatches as £2.60 / 50p, not as a Money dump.
  .param(
    'money',
    /£(?=.*\d.*)[-+]?\d*(?:\.(?=\d.*))?\d*|\d+p/,
    (raw): Money =>
      raw.endsWith('p') ? GBP(Number.parseFloat(raw) / 100) : GBP(Number.parseFloat(raw.slice(1))),
    (m) => (m.value < 1 ? `${Math.round(m.value * 100)}p` : `£${m.value.toFixed(2)}`),
  )
  // The emphasised run IS the parameter: the markers live in the pattern,
  // parse strips them, format restores them. Markup is notation, like £2.50.
  .param(
    'title',
    /\*[^*]+\*/,
    (raw) => raw.slice(1, -1),
    (t) => `*${t}*`,
  )

stimulus('borrowed {title}, due back on {date}', (state, title, due) => ({
  loans: [...state.loans, { title, due }],
}))

stimulus('returns it on {date}', (state, returnedOn) => ({
  fee: state.loans.reduce((fee, loan) => addMoney(fee, lateFee(loan, returnedOn)), GBP(0)),
}))

sensor('owes a {money} late fee', (state) => state.fee)

sensor('{money} for each day overdue', () => FEE_PER_DAY)

stimulus('asks to borrow {title} on {date}', (state, _title, on) => ({
  granted: mayBorrow(state.loans, on),
}))

sensor('the library refuses', (state) => {
  if (state.granted) throw new Error('expected the library to refuse')
})

sensor('the library agrees', (state) => {
  if (!state.granted) throw new Error('expected the library to agree')
})
